runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/net/url/Url.kt

```diff
@@ -0,0 +1,347 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.net.url
+
+import aws.smithy.kotlin.runtime.net.*
+import aws.smithy.kotlin.runtime.text.Scanner
+import aws.smithy.kotlin.runtime.text.encoding.Encodable
+import aws.smithy.kotlin.runtime.text.encoding.PercentEncoding
+import aws.smithy.kotlin.runtime.text.ensurePrefix
+import aws.smithy.kotlin.runtime.util.CanDeepCopy
+
+/**
+ * Represents a full, valid URL
+ * @param scheme The wire protocol (e.g., http, https, etc.)
+ * @param host The [Host] for the URL
+ * @param port The remote port number for the URL (e.g., TCP port)
+ * @param path The path element of this URL
+ * @param parameters The query parameters for this URL.
+ * @param fragment Optional fragment component of this URL (without the `#` character)
+ * @param userInfo Optional user authentication information for this URL
+ */
+public class Url private constructor(
+    public val scheme: Scheme,
+    public val host: Host,
+    public val port: Int,
+    public val path: UrlPath,
+    public val parameters: QueryParameters,
+    public val userInfo: UserInfo,
```

Comment:
**suggestion** this could remain nullable, we don't use it much and unlike `QueryParameters` and `Headers` it's "empty" state is less useful.

Reply:
Now that I've switched this to non-nullable I find I prefer it since it involves less null-coalescing operations for users. I could switch it back if you feel strongly but I have a slight preference for leaving it non-null.